test/extensions/transport_sockets/alts/tsi_socket_test.cc

```diff
@@ -0,0 +1,282 @@
+                   client_.tsi_socket_->doRead(client_.read_buffer_));
+    EXPECT_EQ(makeFakeTsiFrame("CLIENT_FINISHED"), client_to_server_.toString());
+    EXPECT_EQ(0L, client_.read_buffer_.length());
+
+    EXPECT_CALL(*server_.raw_socket_, doRead(_));
+    EXPECT_CALL(*server_.raw_socket_, doWrite(_, false));
+    EXPECT_CALL(server_.callbacks_, raiseEvent(Network::ConnectionEvent::Connected));
+    expectIoResult({Network::PostIoAction::KeepOpen, 0UL, false},
+                   server_.tsi_socket_->doRead(server_.read_buffer_));
+    EXPECT_EQ(makeFakeTsiFrame("SERVER_FINISHED"), server_to_client_.toString());
+
+    EXPECT_CALL(*client_.raw_socket_, doRead(_));
+    EXPECT_CALL(client_.callbacks_, raiseEvent(Network::ConnectionEvent::Connected));
+    expectIoResult({Network::PostIoAction::KeepOpen, 0UL, false},
+                   client_.tsi_socket_->doRead(client_.read_buffer_));
+  }
+
+  struct SocketForTest {
+    std::unique_ptr<TsiSocket> tsi_socket_;
+    NiceMock<Network::MockTransportSocket>* raw_socket_{};
+    NiceMock<Network::MockTransportSocketCallbacks> callbacks_;
+    Buffer::OwnedImpl read_buffer_;
+    Buffer::OwnedImpl write_buffer_;
+  };
+
+  SocketForTest client_;
+  SocketForTest server_;
+
+  Buffer::OwnedImpl client_to_server_;
+  Buffer::OwnedImpl server_to_client_;
+
+  NiceMock<Event::MockDispatcher> dispatcher_;
+};
+
+TEST_F(TsiSocketTest, DoesNotHaveSsl) {
+  initialize(nullptr, nullptr);
+  EXPECT_EQ(nullptr, client_.tsi_socket_->ssl());
+
+  const auto& socket_ = *client_.tsi_socket_;
+  EXPECT_EQ(nullptr, socket_.ssl());
+}
+
+TEST_F(TsiSocketTest, ProxyCallbacks) {
+  initialize(nullptr, nullptr);
+
+  EXPECT_CALL(client_.callbacks_, fd()).WillOnce(Return(111));
+  EXPECT_EQ(111, client_.raw_socket_->callbacks_->fd());
+
+  EXPECT_EQ(&client_.callbacks_.connection_, &client_.raw_socket_->callbacks_->connection());
+
+  EXPECT_FALSE(client_.raw_socket_->callbacks_->shouldDrainReadBuffer());
+}
+
+TEST_F(TsiSocketTest, HandshakeWithoutValidationAndTransferData) {
+  initialize(nullptr, nullptr);
+
+  client_.write_buffer_.add("hello from client");
+
+  doHandshakeAndExpectSuccess();
+  EXPECT_EQ(0L, server_.read_buffer_.length());
+  EXPECT_EQ(0L, client_.read_buffer_.length());
+
+  EXPECT_EQ("", client_.tsi_socket_->protocol());
+
+  EXPECT_CALL(*client_.raw_socket_, doWrite(_, false));
+  expectIoResult({Network::PostIoAction::KeepOpen, 21UL, false},
+                 client_.tsi_socket_->doWrite(client_.write_buffer_, false));
+  EXPECT_EQ(makeFakeTsiFrame("hello from client"), client_to_server_.toString());
+
+  EXPECT_CALL(*server_.raw_socket_, doRead(_));
+  expectIoResult({Network::PostIoAction::KeepOpen, 21UL, false},
+                 server_.tsi_socket_->doRead(server_.read_buffer_));
+  EXPECT_EQ("hello from client", server_.read_buffer_.toString());
+}
+
+TEST_F(TsiSocketTest, HandshakeWithSucessfulValidationAndTransferData) {
```

Comment:
This test looks very similar to the one above. Can you extend doHandshakeAndExpectSuccess() to  successfulHandshakeAndTransferData() so that the shared code between these 2 tests can be in one place?

Reply:
done